### PR TITLE
Add support for reasoning models

### PR DESCRIPTION
**Pull request by RobinPicard**

### CODE REVIEW

File: outlines/backends/__init__.py [hunk 1]
```diff
@@ -56,6 +84,13 @@ def get_json_schema_logits_processor(
         The Outlines model of the user.
     json_schema: str
         The JSON schema to create a logits processor from.
+    end_thinking_tag: str | None
+        The tag the model uses to indicate the end of the thinking process.
+        Only used when running a thinking model.
+    thinking_max_tokens: int | None
+        The maximum number of tokens the model can think about. Only used when
```
Comment (rlouf) [inline outlines/backends/__init__.py:91]:
```suggestion
        The maximum number of tokens the model can think for. Only used when
```

File: outlines/backends/llguidance.py [hunk 2]
```diff
@@ -35,6 +38,10 @@ def __init__(
             The LLGuidance tokenizer
         tensor_library_name: str
             The name of the tensor library used by the model
+        end_thinking_token_id: int | None
+            The id of the end thinking token
+        thinking_max_tokens: int | None
+            The maximum number of tokens the model can think about
```
[the remark this answers is not in the document]
Comment (RobinPicard) [reply outlines/backends/llguidance.py:44]: Initially I wanted to wrap the logits processor into another one that would just not bias anything until we encounter the token and then it calls the other tokenizer it wraps, the problem is that it does not work for batching as the different sequences may not all stop thinking at the same time.
